sdks/java/harness/src/main/java/org/apache/beam/fn/harness/status/BeamFnStatusClient.java

```diff
@@ -0,0 +1,120 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.status;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.StringJoiner;
+import java.util.function.Function;
+import org.apache.beam.fn.harness.control.ProcessBundleHandler.BundleProcessor;
+import org.apache.beam.fn.harness.control.ProcessBundleHandler.BundleProcessorCache;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusRequest;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.WorkerStatusResponse;
+import org.apache.beam.model.fnexecution.v1.BeamFnWorkerStatusGrpc;
+import org.apache.beam.model.pipeline.v1.Endpoints.ApiServiceDescriptor;
+import org.apache.beam.runners.core.metrics.ExecutionStateTracker;
+import org.apache.beam.vendor.grpc.v1p26p0.io.grpc.ManagedChannel;
+import org.apache.beam.vendor.grpc.v1p26p0.io.grpc.stub.StreamObserver;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BeamFnStatusClient {
+  private final StreamObserver<WorkerStatusResponse> outboundObserver;
+  private final BundleProcessorCache processBundleCache;
+  private static final Logger LOG = LoggerFactory.getLogger(BeamFnStatusClient.class);
+
+  public BeamFnStatusClient(
+      ApiServiceDescriptor apiServiceDescriptor,
+      Function<ApiServiceDescriptor, ManagedChannel> channelFactory,
+      BundleProcessorCache processBundleCache) {
+    BeamFnWorkerStatusGrpc.BeamFnWorkerStatusStub stub =
+        BeamFnWorkerStatusGrpc.newStub(channelFactory.apply(apiServiceDescriptor));
+    stub.workerStatus(new InboundObserver());
+    this.outboundObserver = stub.workerStatus(new InboundObserver());
+    this.processBundleCache = processBundleCache;
+  }
+
+  void threadDump(StringJoiner trace) {
+    trace.add("========== THREAD DUMP ==========");
+    Map<Thread, StackTraceElement[]> threads = Thread.getAllStackTraces();
+    threads
+        .entrySet()
+        .forEach(
+            entry -> {
+              trace.add(
+                  String.format(
+                      "%n--- Thread #%d name: %s ---",
+                      entry.getKey().getId(), entry.getKey().getName()));
+              Arrays.stream(entry.getValue()).map(StackTraceElement::toString).forEach(trace::add);
+            });
+  }
+
+  @VisibleForTesting
+  void activeProcessBundleState(StringJoiner activeBundlesState) {
```

Comment:
nit:The meaning of StringJoiner parameter is a little bit confusing to me, maybe just call this getActiveProcessBundleStates() to keep the signature simple and just use the StringJoiner for concating each section at call site.